typescript/provider-sdk/src/module.ts

```diff
@@ -1,20 +1,43 @@
+import { IProvider, ISigner } from './altvm.js';
+
 type AddrMap = Record<string, unknown>;
 export type AnnotatedTx = { annotation?: string; [key: string]: any };
 export type TxReceipt = { [key: string]: any };
 
-export interface HypModuleArgs<Cfg, Addrs extends AddrMap> {
-  addresses: Addrs;
+export interface ModuleType {
+  config: unknown;
+  derived: unknown;
+  addresses: AddrMap;
+}
+
+export type Config<M extends ModuleType> = M['config'];
+export type Derived<M extends ModuleType> = M['derived'];
+export type Addresses<M extends ModuleType> = M['addresses'];
+
+export interface HypModuleArgs<M extends ModuleType> {
+  addresses: Addresses<M>;
   chain: string | number;
-  config: string | Cfg;
+  config: string | Config<M>;
+}
+
+export interface HypReader<M extends ModuleType> {
+  read(address: string): Promise<Derived<M>>;
 }
 
-export interface HypModule<Cfg, Addrs extends AddrMap> {
-  read(): Promise<Cfg>;
-  serialize(): Addrs;
-  update(config: Cfg): Promise<AnnotatedTx[]>;
+export interface HypModule<M extends ModuleType> {
+  read(): Promise<Derived<M>>;
+  serialize(): Addresses<M>;
+  update(config: Config<M>): Promise<AnnotatedTx[]>;
 }
 
-export interface HypModuleFactory<Cfg, Addrs extends AddrMap> {
-  connect(args: HypModuleArgs<Cfg, Addrs>): HypModule<Cfg, Addrs>;
-  create(config: Cfg): Promise<HypModule<Cfg, Addrs>>;
+export interface ModuleProvider<M extends ModuleType> {
+  connectReader: (provider: IProvider) => HypReader<M>;
+  connectModule: (
+    signer: ISigner<AnnotatedTx, TxReceipt>,
+    args: HypModuleArgs<M>,
+  ) => HypModule<M>;
+  createModule: (
+    signer: ISigner<AnnotatedTx, TxReceipt>,
+    config: Config<M>,
+  ) => Promise<HypModule<M>>;
```

Comment:
shouldn't we expose a methods to get an artifact module provider on the `ProtocolProvider`? I see the ism, hook and token provider methods have been removed from the interface

Reply:
Removed because in the AltVM API there is no notion of typed artifacts yet, implementing different kinds of ISMs, hooks, and tokens. So we have a concrete ModuleProvider implementation in `deploy-kit` using the AltVmModules which are generic, but no need yet to expose this here.

I hope we'll reintroduce artifacts and the ArtifactProvider and then reimplement the existing artifact types in Cosmos and Radix.